src/SAWScript/Crucible/LLVM/FFI.hs

```diff
@@ -464,7 +464,21 @@ arrayTypeInfo tenv lenTypes ffiBasicType = do
                 cumulLenTerms = map natOpenTerm $ scanl1 (*) lens
                 arrCryType :| cumulElemTypes =
                   NE.scanr vectorTypeOpenTerm basicCryType lenTerms
-                cumul = zip3 cumulLenTerms (tail lenTerms) (tail cumulElemTypes)
+
+                noArrayLengths :: a
+                noArrayLengths =
+                  panic "arrayTypeInfo"
+                        ["FFIArray with empty list of length types"]
+
+                lenTermsTail =
+                  case lenTerms of
+                    _:lenTermsTail' -> lenTermsTail'
+                    [] -> noArrayLengths
+                cumulElemTypesTail =
+                  case cumulElemTypes of
+                    _:cumulElemTypesTail' -> cumulElemTypesTail'
+                    [] -> noArrayLengths
+                cumul = zip3 cumulLenTerms lenTermsTail cumulElemTypesTail
```

Comment:
Actually, I think that with the way the code is set up, these lists _aren't_ all of the same length. The chain of reasoning is:

* Start with `lenTypes` (which is non-empty by assumption, so length `n + 1`)
* Create `lens` by `map`ping over `lenTypes` (length `n + 1`)
* Create `lenTerms` by `map`ing over `lens` (length `n + 1`)
* Create `cumulLenTerms` by calling `map`/`scanl1` on `lens` (length `n + 1`)
* Create `cumulElemTypes` by calling `scanr` on `lenTerms` (which yields a list of length `n + 2`) and taking the tail of the result (giving `cumulElemTypes` the length `n + 1`)
* Create `lenTermsTail` by taking the tail of `lenTerms` (length `n`)
* Create `cumulElemTypesTail` by taking the tail of `cumulElemTypes` (length `n`)

So `cumulLenTerms` is of length `n + 1`, but `lenTermsTail` and `cumulElemTypesTail` are of length `n`. Note that this doesn't pose any risk of a runtime crash, since `zip3` will always return a list whose size is equal to the minimum length of the three argument lists. Based on the comments near the code below that demonstrates how `cumul` is used, I think this is fine, since you never use the elements at index `n` of `lenTermsTail` or `cumulElemTypesTail`. (I will admit that it does take some squinting before you realize that that is the case, however.)

Reply:
We _could_ rectify this by calling `take (length lenTypes - 1) cumulLenTerms` to ensure that all of the lists are of the same length, but that has runtime performance implications. I am inclined to simply leave a comment here saying that `cumulLenTerms` is of length one greater than the other two argument lists, but that due to how `cumul` is consumed, it is fine that `zip3` discards the last element of `cumulLenTerms`.

Reply:
Yes, I was more concerned with the underlying lists than any unsafeness regarding `zip3` itself, and the comment is a good resolution to my concern.  Thanks!